pkg/parser/parser.go

```diff
@@ -26,27 +26,44 @@
 package parser
 
 import (
+	"fmt"
 	"strings"
+
+	"github.com/raklaptudirm/mash/pkg/command"
+	"github.com/raklaptudirm/mash/pkg/command/builtin"
 )
 
+var NoCommandError = fmt.Errorf("No commands provided")
+
 // Function Parse accepts user command as
 // input and separates the command string
 // from the argument string, and then
 // dispatches the argument string to the
 // Args function. It returns the command
 // and the argument slice.
-func Parse(input string) (string, []string) {
+func Parse(input string) (command.SimpleCommand, error) {
 	input = strings.Trim(input, " \t\n\r")
 	words := Words(input)
 	if len(words) == 0 {
-		return "", []string{}
+		return command.SimpleCommand{}, NoCommandError
 	}
-	return words[0], words[1:]
+	cmd, exists := builtin.Commands[words[0]]
+	if !exists {
+		return command.SimpleCommand{
+			Name: words[0],
+			Args: words[1:],
+		}, nil
+	}
+	cmd.Args = words[1:]
+	return cmd, nil
 }
 
 // Function Args parses an argument string
 // input into an argument slice.
 func Words(input string) []string {
+	// remove leading and trailing whitespaces
+	input = strings.TrimSpace(input)
```

Comment:
This actually done above in the `Parse()` function. You can either remove that one, or this one.

Reply:
Oh I missed that, I will remove that from words.

> Also, I do not understand how this struct will be used. I dont think there is any situation where we will execute a list of simple commands, given that this is not a part of the parser.

For the command struct, the idea is that we usually have many redirections and pipes to and from multiple commands in a single "command line" (until you press return key). And we also have to think about the eventual thing outside of REPL because scripts should also be a part of this right? that means there will be multiple different commands in one execution. If that makes sense. Let me know if you have a different idea though 😄

Reply:
My idea is simple.

The problem with bash, imo, is that the fact each line in it is also a command. I was thinking that mash would work a bit differently. If the first token of a source is a word or a string, mash will treat it as a command, other wise as a script, which  will have different schematics.

The idea is to make commands a datatype in the script, which can then be invoked (a bit like functions, but they will be shell commands). They will be inside some delimiters, like `[]` or `''`, and can be invoked like functions or a special operator `$`.

How does this sound? Do tell me if you have any suggestions.

Reply:
Can you describe what a script in `mash` would look like? Or give a proper example? I don't quite get this part as I thought this was an interpreter REPL which runs a sequence of commands (what bash already does).

Reply:
You can think of it like mash has two types of syntaxes. One will be dominantly used in the repl and the other in scripts(code files). This makes sure when you want to write some shell script in mash, each line is not treated as a command(unlike bash) because it is a script. On the other hand, using the command syntax, the line will be treated as a command and parsed according to my specified grammar. The details of how to determine whether any source is a command or a script, and the schematics of a script will come later, because for now we will be focusing on the command syntax.

Reply:
I personally don't think there should be such distinction between a script and a command. Many times, we have multiple commands in one line and this pattern would create problems for you later. Consider having two syntaxes for Python in script and in prompt, IMO it would be very confusing (and bad because you have two separate types of syntaxes so more overhead for your parser) because the REPL should be representative of the language.

Reply:
Let me think about what we can do.